### PR TITLE
revisions

### PR DESCRIPTION
**Pull request by prjemian**
on discussion

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -29,4 +29,6 @@ repos:
       - id: ruff # replaces Flake8, isort, pydocstyle, pyupgrade
         args:
           - --fix
+          - --ignore=E402
```
Comment (prjemian) [inline .pre-commit-config.yaml:32]: Would this be as effective in this section of `pyproject.toml`?
https://github.com/prjemian/model_instrument/blob/57649763eaa225d859d84f24352d6d5d608ecbf0/pyproject.toml#L187-L189

**Comment by prjemian** [comment]
testing at APS with `ipython -i -c "from instrument.startup import *"`, this exception

<details>

```py
$ from instrument.startup import *
Python 3.12.6 | packaged by conda-forge | (main, Sep 30 2024, 18:08:52) [GCC 13.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.28.0 -- An enhanced Interactive Python. Type '?' for help.
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/epics/ca.py:383, in initialize_libca()
    381     if os.name == 'nt':
    382         load_dll(find_libCom())
--> 383     libca = load_dll(find_libca())
    384 except Exception as exc:
    385     raise ChannelAccessException('loading Epics CA DLL failed: ' + str(exc))

File ~/.conda/envs/model_instrument/lib/python3.12/ctypes/__init__.py:460, in LibraryLoader.LoadLibrary(self, name)
    459 def LoadLibrary(self, name):
--> 460     return self._dlltype(name)

File ~/.conda/envs/model_instrument/lib/python3.12/ctypes/__init__.py:379, in CDLL.__init__(self, name, mode, handle, use_errno, use_last_error, winmode)
    376 self._FuncPtr = _FuncPtr
    378 if handle is None:
--> 379     self._handle = _dlopen(self._name, mode)
    380 else:
    381     self._handle = handle

OSError: libncurses.so.5: cannot open shared object file: No such file or directory
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/epics/ca.py:383, in initialize_libca()
    382         load_dll(find_libCom())
--> 383     libca = load_dll(find_libca())
    384 except Exception as exc:

File ~/.conda/envs/model_instrument/lib/python3.12/ctypes/__init__.py:460, in LibraryLoader.LoadLibrary(self, name)
    459 def LoadLibrary(self, name):
--> 460     return self._dlltype(name)

File ~/.conda/envs/model_instrument/lib/python3.12/ctypes/__init__.py:379, in CDLL.__init__(self, name, mode, handle, use_errno, use_last_error, winmode)
    378 if handle is None:
--> 379     self._handle = _dlopen(self._name, mode)
    380 else:

OSError: libncurses.so.5: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

ChannelAccessException                    Traceback (most recent call last)
Cell In[1], line 1
----> 1 from instrument.startup import *

File /home/beams1/JEMIAN/Documents/projects/prjemian/model_instrument/src/instrument/startup.py:15
     12 # logging setup first
     13 import logging
---> 15 from .utils._logging_setup import configure_logging
     17 configure_logging()
     19 logger = logging.getLogger(__name__)

File /home/beams1/JEMIAN/Documents/projects/prjemian/model_instrument/src/instrument/utils/_logging_setup.py:34
     31 import logging
     32 import pathlib
---> 34 from apstools.utils import file_log_handler
     35 from apstools.utils import setup_IPython_console_logging
     36 from apstools.utils import stream_log_handler

File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/apstools/utils/__init__.py:40
     38 from .catalog import getDefaultDatabase
     39 from .catalog import getStreamValues
---> 40 from .device_info import listdevice
     41 from .email import EmailNotifications
     42 from .image_analysis import analyze_1D

File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/apstools/utils/device_info.py:16
     13 from collections import defaultdict
     15 import pandas as pd
---> 16 from ophyd import Device
     17 from ophyd import Signal
     18 from ophyd.signal import ConnectionTimeoutError

File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/ophyd/__init__.py:89
     85         raise RuntimeError("control layer not set, " "unsure how you got to this state")
     86     return cl
---> 89 set_cl()
     91 # Areadetector-related
     92 from .areadetector import *  # noqa: F401, F402, E402, F403

File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/ophyd/__init__.py:24, in set_cl(control_layer, pv_telemetry)
     22 for c_type in known_layers:
     23     try:
---> 24         set_cl(c_type, pv_telemetry=pv_telemetry)
     25     except ImportError:
     26         continue

File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/ophyd/__init__.py:51, in set_cl(control_layer, pv_telemetry)
     48 else:
     49     raise ValueError("unknown control_layer")
---> 51 shim.setup(logger)
     53 exports = (
     54     "setup",
     55     "caput",
   (...)
     61     "get_dispatcher",
     62 )
     63 # this sets the module level value

File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/ophyd/_pyepics_shim.py:203, in setup(logger)
    199     _dispatcher = None
    201 logger.debug("Installing event dispatcher")
    202 _dispatcher = EventDispatcher(
--> 203     thread_class=PyepicsCallbackThread, context=ca.current_context(), logger=logger
    204 )
    205 atexit.register(_cleanup)
    206 return _dispatcher

File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/epics/ca.py:596, in withCA.<locals>.wrapper(*args, **kwds)
    594 global libca
    595 if libca is None:
--> 596     initialize_libca()
    597 elif libca is _LIBCA_FINALIZED:
    598     return  # Avoid raising exceptions when Python shutting down

File ~/.conda/envs/model_instrument/lib/python3.12/site-packages/epics/ca.py:385, in initialize_libca()
    383     libca = load_dll(find_libca())
    384 except Exception as exc:
--> 385     raise ChannelAccessException('loading Epics CA DLL failed: ' + str(exc))
    387 ca_context = {False:0, True:1}[PREEMPTIVE_CALLBACK]
    388 ret = libca.ca_context_create(ca_context)

ChannelAccessException: loading Epics CA DLL failed: libncurses.so.5: cannot open shared object file: No such file or directory
```

</details>

**Comment by prjemian** [comment]
The root cause is pip has installed incompatible versions of PyEpics and libCom (from EPICS base).  When PyEpics is installed from `conda-forge::pyepics`, this problem is not raised.

**Comment by prjemian** [comment]
Rebuilding the environment with this change, no exception is raised:

```py
$ ipython -i -c "from instrument.startup import *"
Python 3.12.6 | packaged by conda-forge | (main, Sep 30 2024, 18:08:52) [GCC 13.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.28.0 -- An enhanced Interactive Python. Type '?' for help.
Activating auto-logging. Current session state plus future input saved.
Filename       : /home/beams1/JEMIAN/Documents/projects/prjemian/model_instrument/.logs/ipython_console.log
Mode           : rotate
Output logging : True
Raw input log  : False
Timestamping   : True
State          : active
I Fri-09:52:14 - ############################################################ startup
I Fri-09:52:14 - logging started
I Fri-09:52:14 - logging level = 10
I Fri-09:52:14 - /home/beams1/JEMIAN/Documents/projects/prjemian/model_instrument/src/instrument/utils/_logging_setup.py
I Fri-09:52:14 - /home/beams1/JEMIAN/Documents/projects/prjemian/model_instrument/src/instrument/utils/aps_functions.py
I Fri-09:52:14 - /home/beams1/JEMIAN/Documents/projects/prjemian/model_instrument/src/instrument/utils/helper_functions.py
Exception reporting mode: Minimal
I Fri-09:52:14 - xmode exception level: 'Minimal'
W Fri-09:52:16 - APS DM setup file does not exist: '/home/dm/etc/dm.setup.sh'
I Fri-09:52:17 - /home/beams1/JEMIAN/Documents/projects/prjemian/model_instrument/src/instrument/utils/controls_setup.py
I Fri-09:52:17 - /home/beams1/JEMIAN/Documents/projects/prjemian/model_instrument/src/instrument/utils/metadata.py
I Fri-09:52:17 - RunEngine metadata saved in directory: /home/beams/JEMIAN/.config/Bluesky_RunEngine_md
I Fri-09:52:17 - using ophyd control layer: 'pyepics'

In [1]: 

```

**Comment by prjemian** [comment]
In this case, conda must be used for installation of `libCom.so` and other pre-compiled libraries required by PyEpics.

File: docs/source/console.rst [hunk 2]
```diff
@@ -22,34 +22,25 @@ Note that ``In [1]:`` and ``In [2]:`` are numbered command prompts from IPython.
 
     In [1]: from instrument.startup import *
     Activating auto-logging. Current session state plus future input saved.
-    Filename       : /path/to/project/.logs/ipython_console.log
+    Filename       : /home/prjemian/Documents/projects/prjemian/model_instrument/.logs/ipython_console.log
```
Comment (MDecarabas) [inline docs/source/console.rst:25]: why is this path hardcoded?

File: docs/source/console.rst [hunk 3]
```diff
@@ -22,34 +22,25 @@ Note that ``In [1]:`` and ``In [2]:`` are numbered command prompts from IPython.
...
-    I Mon-16:55:40 - /path/to/project/src/instrument/core/epics_setup.py
-    I Mon-16:55:40 - /path/to/project/src/instrument/core/metadata.py
-    I Mon-16:55:40 - RunEngine metadata saved in directory: /home/user/.config/Bluesky_RunEngine_md
-    I Mon-16:55:40 - /path/to/project/src/instrument/core/tests/common.py
+    I Thu-23:46:37 - xmode exception level: 'Minimal'
+    W Thu-23:46:38 - APS DM setup file does not exist: '/home/dm/etc/dm.setup.sh'
+    I Thu-23:46:38 - /home/prjemian/Documents/projects/prjemian/model_instrument/src/instrument/utils/controls_setup.py
+    I Thu-23:46:38 - /home/prjemian/Documents/projects/prjemian/model_instrument/src/instrument/utils/metadata.py
+    I Thu-23:46:38 - RunEngine metadata saved in directory: /home/prjemian/.config/Bluesky_RunEngine_md
+    I Thu-23:46:38 - using ophyd control layer: 'pyepics'
```
Comment (MDecarabas) [inline docs/source/console.rst:43]: not a fan of showing this log dump for documentation but we can remove it on the second revision

File: docs/source/install.rst [hunk 4]
```diff
@@ -35,7 +35,7 @@ environment.
     :linenos:
 
     export INSTALL_ENVIRONMENT_NAME=choose_your_own_name
-    conda create -y -n "${INSTALL_ENVIRONMENT_NAME}" python pyqt=5
+    conda create -y -n "${INSTALL_ENVIRONMENT_NAME}" python pyqt=5 pyepics
```
Comment (MDecarabas) [inline docs/source/install.rst:38]: add python=3.10

**Review by MDecarabas — APPROVED** [review]

File: docs/source/console.rst (hunk 2 above)
Comment (prjemian) [reply docs/source/console.rst:25]: This is output from the session.

File: docs/source/console.rst (hunk 3 above)
Comment (prjemian) [reply docs/source/console.rst:43]: Session output.  Shows the user what to expect.

File: docs/source/install.rst (hunk 4 above)
Comment (prjemian) [reply docs/source/install.rst:38]: Not necessary to pin to _only_ 3.10.  Let the solver and the package requirements decide that.

Comment (MDecarabas) [reply docs/source/install.rst:38]: ok

Comment (prjemian) [reply docs/source/install.rst:38]: For example, this combination works on my APS workstation:

```bash
$ conda list python
# packages in environment at /home/beams/JEMIAN/.conda/envs/model_instrument:
#
# Name                    Version                   Build  Channel
dnspython                 2.6.1                    pypi_0    pypi
ipython                   8.28.0                   pypi_0    pypi
python                    3.12.6          hc5c86c4_2_cpython    conda-forge
python-dateutil           2.9.0.post0              pypi_0    pypi
python-json-logger        2.0.7                    pypi_0    pypi
python-multipart          0.0.12                   pypi_0    pypi
python_abi                3.12                    5_cp312    conda-forge
types-python-dateutil     2.9.0.20241003           pypi_0    pypi
```

File: docs/source/console.rst (hunk 3 above)
Comment (MDecarabas) [reply docs/source/console.rst:43]: To me it seems quite daunting, especially given how much prints to the terminal. As well it makes it so your eyes glaze over what is important vs not. Round 2 I want there to be a less logging default option. No one needs to see that xmode exception level is Minimal

Comment (MDecarabas) [reply docs/source/console.rst:43]: APS DM setup file does not exist: '/home/dm/etc/dm.setup.sh' is an important log, but normally I would have stopped reading long before that

File: docs/source/console.rst (hunk 2 above)
Comment (prjemian) [reply docs/source/console.rst:25]: Requires hand-editing to restore to `path/to/project`

File: docs/source/console.rst (hunk 3 above)
Comment (prjemian) [reply docs/source/console.rst]: Let's shift further comments on logging content to issue #2.

**Pull request merged**